state/postgres_statement.go

```diff
@@ -104,3 +104,11 @@ func (stmt DiffedPostgresStatementStats) Add(other DiffedPostgresStatementStats)
 		BlkWriteTime:      stmt.BlkWriteTime + other.BlkWriteTime,
 	}
 }
+
+type QueryIdentity struct {
+	QueryID int64
+	Fingerprint uint64
+	LastSeen time.Time
+}
+
+type QueryIdentityMap map[int64]QueryIdentity
```

Comment:
It looks like query IDs are globally unique, since the relation ID of referenced tables is used when generating the fingerprint:

https://github.com/postgres/postgres/blob/fc7852c6cb89a5384e0b4ad30874de92f63f88be/src/backend/utils/misc/queryjumble.c#L285

The question is: are duplicate query IDs across databases (that were for entirely different queries) a big enough issue that we need to track which database the query ID came from in `QueryIdentityMap`?

Reply:
Checking pg_stat_statements in our own database, it looks like we have ~0.15% dupes. That's pretty small, but maybe not small enough that we can just ignore it. In our case, though, all the dupes are in one database. What's the impact of collisions here? We may associate stats with the incorrect query text?

Reply:
Yes, the stats could end up being associated with the wrong query in that case. But we only fall back to this identity map for queries that have churned out of `pg_stat_statements` since the last full snapshot so it may not be a big deal.

Reply:
So you think altogether it's unlikely enough that we can just ignore it? I'm a little reluctant to go with that (because if this does bite you, it will be basically impossible to figure out), but given the odds, it doesn't seem unreasonable, and I don't have a great idea on how to surface the problem. (And I don't think it's worth talking about it as a potential problem, since it's so unlikely).

Reply:
Trying to think through this:

1) Queries on two different databases having the same queryid, e.g. `SELECT 1` would have the same queryid, despite it being on different databases (since it doesn't involve any table OID that'd be database specific)

2) A hash collision with the Postgres `queryid`, i.e. two entirely unrelated queries in two different databases having the same queryid (if it was the same database we wouldn't know about it, since they'd just end up being a single entry in pgss)

If I understand the conversation so far, you're talking about (2) here, though I'm not sure if the 0.15% statistic that @msakrejda referenced is for actual hash collisions? (I'd be surprised, I'm assuming its more likely these are actually identical queries).

Assuming the hash collision case (2), what would happen today:

1. database A has queryid 1, and querytext X
2. database B also has queryid 1, but querytext Y
3. we loop through the stat statements entries, fingerprinting each based on the querytext, and end up with two distinct query fingerprints (and thus associate to the correct query entry in pganalyze)

Now assuming that in a future iteration of this PR we actually check the cache first and don't fingerprint if needed, the problem here would be that we are now associating to the wrong query in the case of the collision.

I'm 50/50 on whether this is a problem in practice. Hash collisions could occur (and the standard Postgres hash function used in pgss isn't the best hash function either), but it does seem a quite unlikely edge case, and we are not in a security sensitive context here, the way I look at it.

The one thing I would be worried about from a security perspective is in case we consider caching query text here as well (or if this mapping is relied on elsewhere to get the text), because doing a simple mapping here could cause query text to cross a security boundary (in the pganalyze app you may be restricted to viewing a single database on a server).

Maybe its better to just use the full triplet (i.e. `(database_id, role_id, queryid)`) for the mapping here, to avoid any accidental errors?

Reply:
>Now assuming that in a future iteration of this PR we actually check the cache first and don't fingerprint if needed, the problem here would be that we are now associating to the wrong query in the case of the collision.
>
>I'm 50/50 on whether this is a problem in practice. Hash collisions could occur (and the standard Postgres hash function used in pgss isn't the best hash function either), but it does seem a quite unlikely edge case, and we are not in a security sensitive context here, the way I look at it.

I implemented that in a followup PR: #346. I don't think hash collisions in this case are likely enough to warrant a more complex implementation at this point in time.

>Maybe its better to just use the full triplet (i.e. (database_id, role_id, queryid)) for the mapping here, to avoid any accidental errors?

FWIW it looks like composite key maps aren't as fast as we'd expect (from the #346 screenshot):

<img width="1356" alt="Screenshot 2022-12-16 at 3 29 04 PM" src="https://user-images.githubusercontent.com/688886/208192118-9d6f1601-7344-47ee-b3a1-22f061dea100.png">

Reply:
Can you run a more isolated benchmark on the performance difference here?

I think the added complexity isn't an issue, since using the full triplet would match what we're doing elsewhere with `PostgresStatementKey` -- and doing it differently actually makes it more complex because now we have to remember when to use a simple queryId and when to use the full `PostgresStatementKey` (and as mentioned we can't just change things to always use `queryId` as the key).

Reply:
Changing to a composite key would necessarily mean increasing the cache size limit. The collector already has an unresolved memory consumption issue, and I worry that increasing the cache size further is going to make that worse.

Reply:
I'll work on building a standalone benchmark script when I'm back from holidays. There might be an argument to drop `PostgresStatementKey` across the board in favor of the query ID on its own.

Reply:
> Changing to a composite key would necessarily mean increasing the cache size limit.

With cache size limit you mean the 100k entries limit on the map?

I'm not sure if that'd need to be raised if the key also includes the database ID and role ID - the overall number of entries shouldn't actually increase much, unless we have hash collisions (which would no longer be the same entry), or the same query without table references runs on multiple databases.

Adding the role ID on the other hand may have such an effect (e.g. imagine a workload using short-lived roles), so there may be an argument for only doing `(database_id, queryid)` without the role ID.

> There might be an argument to drop PostgresStatementKey across the board in favor of the query ID on its own.

I don't think we can drop it across the board because of the aforementioned issue with the same queryid having different text on different databases, and that being security-relevant (e.g. imagine db1 is for tenant1, and has a query comment like `SELECT 1 /* user.email: tenant1@example.com */` and db2 has `SELECT 1 /* user.email: tenant2@example.com */`)

Therefore we'd need to keep the `PostgresStatementKey` use for any maps mapping to query texts at the very least, and anything related to statistics is also problematic (since there you need it to count each database separately).

Reply:
>If I understand the conversation so far, you're talking about (2) here, though I'm not sure if the 0.15% statistic that @msakrejda referenced is for actual hash collisions? (I'd be surprised, I'm assuming its more likely these are actually identical queries).

They were queries with identical queryids occurring multiple times in `pg_stat_statements`. However, investigating further, it looks like this is all due to different records across separate `userid`s in `pg_stat_statements`. If I filter down to a single user, we have no collisions.

Reply:
> Therefore we'd need to keep the `PostgresStatementKey` use for any maps mapping to query texts at the very least, and anything related to statistics is also problematic (since there you need it to count each database separately).

Good point.

Am I misremembering, or had we previously discussed no longer distinguishing between user roles for queries, since it creates a confusing user experience? If so we could improve the map hashing time by using an array/slice (that's indexed into based on `DatabaseIdx`) which itself contains a map of the query ID, and whatever data is being stored.

https://gist.github.com/seanlinsley/dd7b2bf8d09b6ba710d27044794f86c5#file-map_test-go-L182

<img width="565" alt="Screenshot 2022-12-28 at 11 41 50 AM" src="https://user-images.githubusercontent.com/688886/209851500-d7069f82-72fc-4dc3-8fc4-d667e31578bd.png">

But that's all a tangent. For this PR specifically, do you think `QueryIdentityMap` needs to be scoped per database?